### PR TITLE
Module 3 - Asynchronous ASP.NET Core APIs

### PR DESCRIPTION
**Pull request by Kirill-Ovc**

### CODE REVIEW

File: EventTicketingSystem.API/Controllers/EventsController.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        [HttpGet("{cityId}")]
+        public async Task<IActionResult> GetByCity(int cityId)
+        {
+            if (cityId == default)
+            {
+                return BadRequest("City ID is required");
+            }
+            var events = await _eventService.GetEventsByCity(cityId);
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/EventsController.cs]: empty line after the if block

File: EventTicketingSystem.API/Controllers/EventsController.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        public async Task<IActionResult> Get()
+        {
+            var events = await _eventService.GetAllEvents();
+            return Ok(events);
+        }
+
+        [HttpGet("{cityId}")]
+        public async Task<IActionResult> GetByCity(int cityId)
+        {
+            if (cityId == default)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/EventsController.cs]: better to use nullable int

File: EventTicketingSystem.API/Controllers/EventsController.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                return BadRequest("City ID is required");
+            }
+            var events = await _eventService.GetEventsByCity(cityId);
+            return Ok(events);
+        }
+
+        [HttpGet("events/{eventId}/sections/{sectionId}/seats")]
+        public async Task<IActionResult> GetSeats(int eventId, int sectionId)
+        {
+            if (eventId == default || sectionId == default)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/EventsController.cs]: same for both

File: EventTicketingSystem.API/Controllers/IdentityController.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    public class IdentityController : ControllerBase
+    {
+        private readonly SymmetricSecurityKey _key;
+        private static readonly TimeSpan _tokenLifetime = TimeSpan.FromHours(10);
+        private const string Issuer = "localhost";
+        private const string Audience = "localhost";
+
+        public IdentityController(IConfiguration configuration)
+        {
+            var tokenKey = configuration["JwtSettings:SecretKey"];
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/IdentityController.cs]: Please use IOptions<> for work with configuration

File: EventTicketingSystem.API/Controllers/IdentityController.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+using System.Text;
+
+namespace EventTicketingSystem.API.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class IdentityController : ControllerBase
+    {
+        private readonly SymmetricSecurityKey _key;
+        private static readonly TimeSpan _tokenLifetime = TimeSpan.FromHours(10);
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/IdentityController.cs]: make sense to have configurable?

File: EventTicketingSystem.API/Controllers/OrdersController.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        public OrdersController(IOrderService orderService)
+        {
+            _orderService = orderService;
+        }
+
+        [HttpGet("carts/{cartId}")]
+        public async Task<ActionResult<Cart>> GetCart(string cartId)
+        {
+            var cart = await _orderService.GetCart(cartId);
+            if (cart == null)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/OrdersController.cs]: is null + tenar operator. Also looks like exception + error filter might be better solution to remove business logic from controller

File: EventTicketingSystem.API/Controllers/OrdersController.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            return Ok(cart);
+        }
+
+        [HttpPost("carts/{cartId}")]
+        public async Task<ActionResult<Cart>> AddToCart(string cartId, SeatOrder order)
+        {
+            try
+            {
+                var cart = await _orderService.AddToCart(cartId, order);
+                if (cart == null)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/OrdersController.cs]: same + empty line after the if block

File: EventTicketingSystem.API/Controllers/OrdersController.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                {
+                    return NotFound();
+                }
+                return Ok(cart);
+            }
+            catch (InvalidOperationException e)
+            {
+                return BadRequest(e.Message);
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/OrdersController.cs]: 2 cases of same catch: please use exception filter for such translations

File: EventTicketingSystem.API/Controllers/OrdersController.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            }
+        }
+
+        [HttpDelete("carts/{cartId}/events/{eventId}/seats/{seatId}")]
+        public async Task<ActionResult<Cart>> DeleteSeat(string cartId, int eventId, int seatId)
+        {
+            var cart = await _orderService.RemoveFromCart(cartId, seatId);
+            if (cart == null)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/OrdersController.cs]: is null

File: EventTicketingSystem.API/Controllers/OrdersController.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+        [HttpPut("carts/{cartId}/book")]
+        public async Task<ActionResult<int>> BookCart(string cartId)
+        {
+            try
+            {
+                var paymentId = await _orderService.CheckoutCart(cartId);
+                return Ok(paymentId);
+            }
+            catch (InvalidOperationException e)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/OrdersController.cs]: exception filter

File: EventTicketingSystem.API/Controllers/PaymentsController.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+        public PaymentsController(IPaymentService paymentService)
+        {
+            _paymentService = paymentService;
+        }
+
+        [HttpGet("{paymentId}")]
+        public async Task<ActionResult<string>> GetStatus(int paymentId)
+        {
+            try
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/PaymentsController.cs]: same for all comments

File: EventTicketingSystem.API/Controllers/UsersController.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+        public UsersController(IUserService userService)
+        {
+            _userService = userService;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Post([FromBody] UserInfoDto user)
+        {
+            if (user == null)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/UsersController.cs]: same

File: EventTicketingSystem.API/Controllers/VenuesController.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        public async Task<IActionResult> Get()
+        {
+            var users = await _venueRepository.GetVenues();
+            return Ok(users);
+        }
+
+        [HttpGet("{cityId}")]
+        public async Task<IActionResult> GetByCity(int cityId)
+        {
+            if (cityId == default)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Controllers/VenuesController.cs]: same

File: EventTicketingSystem.API/Models/Cart.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
+﻿using EventTicketingSystem.DataAccess.Models.Enums;
+
+namespace EventTicketingSystem.API.Models
+{
+    public class Cart
+    {
+        public string CartId { get; set; }
```
Comment (D9bolic) [inline EventTicketingSystem.API/Models/Cart.cs:5]: empty line between methods/props everywhere

File: EventTicketingSystem.API/Services/OrderService.cs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+            _logger = logger;
+            _bookingRepository = bookingRepository;
+            _bookingSeatRepository = bookingSeatRepository;
+            _paymentRepository = paymentRepository;
+        }
+
+        public async Task<Cart> GetCart(string cartId)
+        {
+            var booking = await _bookingRepository.GetByUuid(cartId);
+            if (booking == null)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/OrderService.cs]: is null

File: EventTicketingSystem.API/Services/OrderService.cs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+            {
+                _logger.LogWarning("Cart not found. CartId = {CartId}", cartId);
+                return null;
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/OrderService.cs]: please avoid returning nulls, better to use custom exceptions for such cases

File: EventTicketingSystem.API/Services/OrderService.cs [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+            _bookingSeatRepository = bookingSeatRepository;
+            _paymentRepository = paymentRepository;
+        }
+
+        public async Task<Cart> GetCart(string cartId)
+        {
+            var booking = await _bookingRepository.GetByUuid(cartId);
+            if (booking == null)
+            {
+                _logger.LogWarning("Cart not found. CartId = {CartId}", cartId);
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/OrderService.cs]: +1

File: EventTicketingSystem.API/Services/OrderService.cs [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+            var booking = await _bookingRepository.GetByUuid(cartId);
+            if (booking == null)
+            {
+                booking = await CreateBooking(order.UserId, cartId);
+            }
+
+            var cart = MapBookingToCart(booking);
+
+            var bookingSeat = await _bookingSeatRepository.GetSeat(order.EventSeatId);
+            if (bookingSeat != null && bookingSeat.BookingId != booking.Id)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/OrderService.cs]: is not null + I believe pattern matching can be used here

File: EventTicketingSystem.API/Services/PaymentService.cs [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+            _bookingRepository = bookingRepository;
+            _bookingSeatRepository = bookingSeatRepository;
+        }
+
+        public async Task<PaymentStatus> GetPaymentStatus(int paymentId)
+        {
+            var payment = await _paymentRepository.Find(paymentId);
+            if (payment == null)
+            {
+                throw new InvalidOperationException("Payment not found");
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/PaymentService.cs]: better to use custom bussines exception here/ Smth like payment not found exception, that inherit from InvalidOperationException

File: EventTicketingSystem.API/Services/PaymentService.cs [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+            var booking = await _bookingRepository.Find(payment.BookingId);
+            if (booking is { Status: BookingStatus.Active } && booking.Price == payment.Amount)
+            {
+                booking.Status = status;
+                await _bookingRepository.Update(booking);
+                await _bookingSeatRepository.UpdateSeatsStatus(booking.Id, seatStatus);
+            }
+            else if (booking is not { Status: BookingStatus.Active })
+            {
+                throw new InvalidOperationException("No active booking");
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/PaymentService.cs]: same about custom exception

File: EventTicketingSystem.API/Services/PaymentService.cs [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+            await UpdateBookingStatus(payment, BookingStatus.Cancelled, EventSeatStatus.Available);
+
+            return payment.PaymentStatus;
+        }
+
+        private async Task UpdateBookingStatus(Payment payment, BookingStatus status, EventSeatStatus seatStatus)
+        {
+            var booking = await _bookingRepository.Find(payment.BookingId);
+            if (booking is { Status: BookingStatus.Active } && booking.Price == payment.Amount)
```
Comment (D9bolic) [inline EventTicketingSystem.API/Services/PaymentService.cs]: make sense to migrate on switch + when?

File: EventTicketingSystem.DataAccess/Services/BookingSeatRepository.cs [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+            return await _context.BookingSeats
+                .Include(bs => bs.EventSeat)
+                .ThenInclude(es => es.Event)
+                .Where(bs => bs.BookingId == bookingId)
+                .ToListAsync();
+        }
+
+        public async Task AddSeat(int bookingId, int eventSeatId, int offerId)
+        {
+            var offer = await _context.Offers.FindAsync(offerId);
```
Comment (D9bolic) [inline EventTicketingSystem.DataAccess/Services/BookingSeatRepository.cs]: looks like  bussines logic as part of data access only

**Review by D9bolic — CHANGES_REQUESTED** [review]

File: EventTicketingSystem.API/Controllers/EventsController.cs (hunk 1 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/EventsController.cs]: Done

File: EventTicketingSystem.API/Controllers/EventsController.cs (hunk 2 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/EventsController.cs]: Fixed

File: EventTicketingSystem.API/Controllers/EventsController.cs (hunk 3 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/EventsController.cs]: Done

File: EventTicketingSystem.API/Controllers/IdentityController.cs (hunk 4 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/IdentityController.cs]: Done

File: EventTicketingSystem.API/Controllers/IdentityController.cs (hunk 5 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/IdentityController.cs]: Done

File: EventTicketingSystem.API/Controllers/OrdersController.cs (hunk 6 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/OrdersController.cs]: Fixed

File: EventTicketingSystem.API/Controllers/OrdersController.cs (hunk 7 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/OrdersController.cs]: Fixed

File: EventTicketingSystem.API/Controllers/OrdersController.cs (hunk 8 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/OrdersController.cs]: Fixed

File: EventTicketingSystem.API/Controllers/OrdersController.cs (hunk 9 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/OrdersController.cs]: Fixed

File: EventTicketingSystem.API/Controllers/OrdersController.cs (hunk 10 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/OrdersController.cs]: Fixed

File: EventTicketingSystem.API/Controllers/PaymentsController.cs (hunk 11 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/PaymentsController.cs]: Done

File: EventTicketingSystem.API/Controllers/UsersController.cs (hunk 12 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/UsersController.cs]: Done

File: EventTicketingSystem.API/Controllers/VenuesController.cs (hunk 13 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Controllers/VenuesController.cs]: Done

File: EventTicketingSystem.API/Services/OrderService.cs (hunk 15 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Services/OrderService.cs]: Fixed

File: EventTicketingSystem.API/Services/OrderService.cs (hunk 16 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Services/OrderService.cs]: Fixed

File: EventTicketingSystem.API/Services/OrderService.cs (hunk 18 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Services/OrderService.cs]: Changed, but pattern matching can't be used here

File: EventTicketingSystem.API/Services/PaymentService.cs (hunk 19 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Services/PaymentService.cs]: Done

File: EventTicketingSystem.API/Services/PaymentService.cs (hunk 21 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Services/PaymentService.cs]: Done

File: EventTicketingSystem.API/Services/PaymentService.cs (hunk 20 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.API/Services/PaymentService.cs]: Done

File: EventTicketingSystem.DataAccess/Services/BookingSeatRepository.cs (hunk 22 above)
Comment (Kirill-Ovc) [reply EventTicketingSystem.DataAccess/Services/BookingSeatRepository.cs]: Done. The method moved to the OrderService.

**Pull request merged**